chain-signatures/node/src/protocol/signature.rs

```diff
@@ -160,16 +162,60 @@ impl SignQueue {
         &self,
         threshold: usize,
         stable: &BTreeSet<Participant>,
+        participants: &Participants,
         indexed: IndexedSignRequest,
-        reorganize: bool,
+        initial_round: usize,
     ) -> SignRequest {
         let sign_id = indexed.id;
+        let reorganize = initial_round > 0;
+        let mut participants = participants.keys_vec();
+        participants.sort();
+
+        // Simple round-robin selection of the proposer, using only inputs that
+        // are guaranteed to be the same on all nodes.
+        fn proposer_per_round(
+            round: usize,
+            participants: &[Participant],
+            entropy: &[u8; 32],
+        ) -> Participant {
+            // if entropy is random, using one byte is as good as using all
+            let index = entropy[0] as usize + round;
+            participants[index % participants.len()]
+        }
+
+        let max_rounds = initial_round + 512;
+        // Use the smallest round that selects a stable proposer.
+        let (round, proposer) = (initial_round..max_rounds)
+            .map(|round| {
+                (
+                    round,
+                    proposer_per_round(round, &participants, &indexed.args.entropy),
+                )
+            })
+            .find(|(_, potential_proposer)| stable.contains(potential_proposer))
+            // on exhausting all rounds, just pick one at random and have posits error out.
+            .unwrap_or_else(|| {
+                (
+                    max_rounds,
+                    *stable
+                        .iter()
+                        .choose(&mut StdRng::from_seed(indexed.args.entropy))
+                        .unwrap(),
+                )
+            });
+
         // NOTE: reorganize, will use the same entropy for reorganizing the participants. The only
         // thing that would be different is the passed in stable participants.
         let mut rng = StdRng::from_seed(indexed.args.entropy);
-        let subset = stable.iter().copied().choose_multiple(&mut rng, threshold);
+        let mut subset = stable.iter().copied().choose_multiple(&mut rng, threshold);
+        if !subset.contains(&proposer) {
+            // replace the last randomly chosen participant with the proposer
+            subset.pop();
+            subset.push(proposer);
+        }
+        subset.sort();
```

Comment:
If we still select a random subset from `stable` on every node, this is going to be very unstable. Nodes will disagree on the subset quite often if even just a single peer has intermittent disconnects.

This should only run on the proposer. Other nodes must use the set provided by the proposer.

Reply:
sounds good, I will get rid of it. Will be using posit participants intersected with only presignature participants in this case then